File: go/connector/sync_test.go
// SPDX-License-Identifier: Apache-2.0

package connector_test

import (
	"context"
	"testing"

	"github.com/jeffs-brain/memory/go/connector"
)

func TestSyncStateManager_SetAndGet(t *testing.T) {
	store := newMemStore()
	mgr := connector.NewSyncStateManager(store)
	ctx := context.Background()

	cursor := connector.SyncCursor{
		Value:    "ts:1234567890.123456",
		Metadata: map[string]string{"channel": "C123"},
	}

	if err := mgr.SetCursor(ctx, "slack", "brain-1", cursor); err != nil {
		t.Fatalf("SetCursor: %v", err)
	}

	loaded, found, err := mgr.GetCursor(ctx, "slack", "brain-1")
	if err != nil {
		t.Fatalf("GetCursor: %v", err)
	}
	if !found {
		t.Fatal("GetCursor returned found=false after SetCursor")
	}
	if loaded.Value != "ts:1234567890.123456" {
		t.Errorf("Value = %q, want %q", loaded.Value, "ts:1234567890.123456")
	}
	if loaded.UpdatedAt.IsZero() {
		t.Error("UpdatedAt should not be zero")
	}
	if loaded.Metadata["channel"] != "C123" {
		t.Errorf("Metadata[channel] = %q, want %q", loaded.Metadata["channel"], "C123")
	}
}

func TestSyncStateManager_GetNonExistent(t *testing.T) {
	store := newMemStore()
	mgr := connector.NewSyncStateManager(store)
	ctx := context.Background()

	_, found, err := mgr.GetCursor(ctx, "unknown", "brain-1")
	if err != nil {
		t.Fatalf("GetCursor: %v", err)
	}
	if found {
		t.Error("GetCursor returned found=true for nonexistent cursor")
	}
}

func TestSyncStateManager_ClearCursor(t *testing.T) {
	store := newMemStore()
	mgr := connector.NewSyncStateManager(store)
	ctx := context.Background()

	cursor := connector.SyncCursor{Value: "cursor-abc"}
	if err := mgr.SetCursor(ctx, "slack", "brain-1", cursor); err != nil {
		t.Fatalf("SetCursor: %v", err)
	}

	if err := mgr.ClearCursor(ctx, "slack", "brain-1"); err != nil {
		t.Fatalf("ClearCursor: %v", err)
	}

	_, found, err := mgr.GetCursor(ctx, "slack", "brain-1")
	if err != nil {
		t.Fatalf("GetCursor after clear: %v", err)
	}
	if found {
		t.Error("cursor found after ClearCursor")
	}
}

func TestSyncStateManager_MultiConnectorIsolation(t *testing.T) {
	store := newMemStore()
	mgr := connector.NewSyncStateManager(store)
	ctx := context.Background()

	slackCursor := connector.SyncCursor{Value: "slack-cursor"}
	gdriveCursor := connector.SyncCursor{Value: "gdrive-cursor"}

	if err := mgr.SetCursor(ctx, "slack", "brain-1", slackCursor); err != nil {
		t.Fatalf("SetCursor slack: %v", err)
	}
	if err := mgr.SetCursor(ctx, "gdrive", "brain-1", gdriveCursor); err != nil {
		t.Fatalf("SetCursor gdrive: %v", err)
	}

	loadedSlack, foundSlack, err := mgr.GetCursor(ctx, "slack", "brain-1")
	if err != nil {
		t.Fatalf("GetCursor slack: %v", err)
	}
	if !foundSlack || loadedSlack.Value != "slack-cursor" {
		t.Errorf("slack cursor: found=%v, value=%q", foundSlack, loadedSlack.Value)
	}

	loadedGdrive, foundGdrive, err := mgr.GetCursor(ctx, "gdrive", "brain-1")
	if err != nil {
		t.Fatalf("GetCursor gdrive: %v", err)
	}
	if !foundGdrive || loadedGdrive.Value != "gdrive-cursor" {
		t.Errorf("gdrive cursor: found=%v, value=%q", foundGdrive, loadedGdrive.Value)
	}
}

func TestSyncStateManager_MultiBrainIsolation(t *testing.T) {
	store := newMemStore()
	mgr := connector.NewSyncStateManager(store)
	ctx := context.Background()

	cursorA := connector.SyncCursor{Value: "cursor-brain-a"}
	cursorB := connector.SyncCursor{Value: "cursor-brain-b"}

	if err := mgr.SetCursor(ctx, "slack", "brain-a", cursorA); err != nil {
		t.Fatalf("SetCursor brain-a: %v", err)
	}
	if err := mgr.SetCursor(ctx, "slack", "brain-b", cursorB); err != nil {
		t.Fatalf("SetCursor brain-b: %v", err)
	}

	loadedA, foundA, err := mgr.GetCursor(ctx, "slack", "brain-a")
	if err != nil {
		t.Fatalf("GetCursor brain-a: %v", err)
	}
	if !foundA || loadedA.Value != "cursor-brain-a" {
		t.Errorf("brain-a cursor: found=%v, value=%q", foundA, loadedA.Value)
	}

	loadedB, foundB, err := mgr.GetCursor(ctx, "slack", "brain-b")
	if err != nil {
		t.Fatalf("GetCursor brain-b: %v", err)
	}
	if !foundB || loadedB.Value != "cursor-brain-b" {
		t.Errorf("brain-b cursor: found=%v, value=%q", foundB, loadedB.Value)
	}
}

func TestSyncStateManager_GenerationIncrement(t *testing.T) {
	store := newMemStore()
	mgr := connector.NewSyncStateManager(store)
	ctx := context.Background()

	// Write twice and verify the cursor is updated.
	cursor1 := connector.SyncCursor{Value: "v1"}
	if err := mgr.SetCursor(ctx, "slack", "brain-1", cursor1); err != nil {
		t.Fatalf("SetCursor v1: %v", err)
	}

	cursor2 := connector.SyncCursor{Value: "v2"}
	if err := mgr.SetCursor(ctx, "slack", "brain-1", cursor2); err != nil {
		t.Fatalf("SetCursor v2: %v", err)
	}

	loaded, found, err := mgr.GetCursor(ctx, "slack", "brain-1")
	if err != nil {
		t.Fatalf("GetCursor: %v", err)
	}
	if !found || loaded.Value != "v2" {
		t.Errorf("cursor: found=%v, value=%q, want v2", found, loaded.Value)
	}
}

func TestSyncStateManager_ClearNonExistent(t *testing.T) {
	store := newMemStore()
	mgr := connector.NewSyncStateManager(store)
	ctx := context.Background()

	// Clearing a nonexistent cursor should not return an error.
	if err := mgr.ClearCursor(ctx, "nonexistent", "brain-1"); err != nil {
		t.Fatalf("ClearCursor nonexistent: %v", err)
	}
}

File: sdks/ts/memory/src/connector/sync.test.ts
// SPDX-License-Identifier: Apache-2.0

import { describe, expect, it } from 'vitest'
import { createMemStore } from '../store/index.js'
import { SyncStateManager } from './sync.js'

describe('SyncStateManager', () => {
  it('sets and gets a cursor', async () => {
    const store = createMemStore()
    const mgr = new SyncStateManager(store)

    await mgr.setCursor('slack', 'brain-1', {
      value: 'ts:1234567890.123456',
      updatedAt: new Date(),
      metadata: { channel: 'C123' },
    })

    const cursor = await mgr.getCursor('slack', 'brain-1')
    expect(cursor).toBeDefined()
    expect(cursor!.value).toBe('ts:1234567890.123456')
    expect(cursor!.updatedAt).toBeInstanceOf(Date)
    expect(cursor!.metadata).toEqual({ channel: 'C123' })
  })

  it('returns undefined for nonexistent cursor', async () => {
    const store = createMemStore()
    const mgr = new SyncStateManager(store)

    const cursor = await mgr.getCursor('unknown', 'brain-1')
    expect(cursor).toBeUndefined()
  })

  it('clears a cursor', async () => {
    const store = createMemStore()
    const mgr = new SyncStateManager(store)

    await mgr.setCursor('slack', 'brain-1', {
      value: 'cursor-abc',
      updatedAt: new Date(),
    })

    await mgr.clearCursor('slack', 'brain-1')
    const cursor = await mgr.getCursor('slack', 'brain-1')
    expect(cursor).toBeUndefined()
  })

  it('isolates cursors across connectors', async () => {
    const store = createMemStore()
    const mgr = new SyncStateManager(store)

    await mgr.setCursor('slack', 'brain-1', {
      value: 'slack-cursor',
      updatedAt: new Date(),
    })
    await mgr.setCursor('gdrive', 'brain-1', {
      value: 'gdrive-cursor',
      updatedAt: new Date(),
    })

    const slack = await mgr.getCursor('slack', 'brain-1')
    const gdrive = await mgr.getCursor('gdrive', 'brain-1')

    expect(slack!.value).toBe('slack-cursor')
    expect(gdrive!.value).toBe('gdrive-cursor')
  })

  it('isolates cursors across brains', async () => {
    const store = createMemStore()
    const mgr = new SyncStateManager(store)

    await mgr.setCursor('slack', 'brain-a', {
      value: 'cursor-a',
      updatedAt: new Date(),
    })
    await mgr.setCursor('slack', 'brain-b', {
      value: 'cursor-b',
      updatedAt: new Date(),
    })

    const a = await mgr.getCursor('slack', 'brain-a')
    const b = await mgr.getCursor('slack', 'brain-b')

    expect(a!.value).toBe('cursor-a')
    expect(b!.value).toBe('cursor-b')
  })

  it('overwrites cursor with generation increment', async () => {
    const store = createMemStore()
    const mgr = new SyncStateManager(store)

    await mgr.setCursor('slack', 'brain-1', {
      value: 'v1',
      updatedAt: new Date(),
    })
    await mgr.setCursor('slack', 'brain-1', {
      value: 'v2',
      updatedAt: new Date(),
    })

    const cursor = await mgr.getCursor('slack', 'brain-1')
    expect(cursor!.value).toBe('v2')
  })

  it('clear is no-op for nonexistent cursor', async () => {
    const store = createMemStore()
    const mgr = new SyncStateManager(store)

    await expect(mgr.clearCursor('nonexistent', 'brain-1')).resolves.toBeUndefined()
  })
})
